cmd/cspc-operator/app/cleanup.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2020 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	bdc "github.com/openebs/maya/pkg/blockdeviceclaim/v1alpha1"
+	apiscspc "github.com/openebs/maya/pkg/cstor/poolcluster/v1alpha1"
+	cspi "github.com/openebs/maya/pkg/cstor/poolinstance/v1alpha3"
+	util "github.com/openebs/maya/pkg/util"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+)
+
+// cleanupCSPIResources removes the CSPI resources when a CSPI is
+// deleted or downscaled
+func cleanupCSPIResources(cspcObj *apis.CStorPoolCluster) error {
+	cspiList, err := cspi.NewKubeClient().WithNamespace(cspcObj.Namespace).List(
+		metav1.ListOptions{
+			LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + cspcObj.Name,
+		},
+	)
+	if err != nil {
+		return errors.Errorf("failed to list cspi for cspc %s to perform cleanup: %s", cspcObj.Name, err.Error())
+	}
+	opts := []cspiCleanupOptions{cleanupBDC}
+	for _, cspiItem := range cspiList.Items {
+		cspiItem := cspiItem // pin it
+		cspiObj := &cspiItem
+		if cspiObj.DeletionTimestamp != nil {
+			// if PoolProtectionFinalizer is not removed wait for the next reconcile
+			// attempt to perform cleanup
+			if canPerformCSPICleanup(cspiObj) {
+				for _, o := range opts {
+					err = o(cspiObj)
+					if err != nil {
+						return errors.Wrapf(err, "failed to cleanup cspi %s for cspc %s", cspiItem.Name, cspcObj.Name)
+					}
+				}
+				cspiObj.Finalizers = util.RemoveString(cspiObj.Finalizers, apiscspc.CSPCFinalizer)
+				_, err = cspi.NewKubeClient().WithNamespace(cspiItem.Namespace).Update(cspiObj)
+				if err != nil {
+					return errors.Wrapf(err, "failed to remove finalizer from cspi %s", cspiItem.Name)
+				}
+				klog.Infof("cleanup for cspi %s was successful", cspiItem.Name)
+			} else {
+				// returning error helps prevent removal of finalizer on cspc object
+				// cspc object should not get deleted before all cspi are deleted successfully
+				return errors.Errorf("failed to cleanup cspi %s for cspc %s : waiting for pool to get destroyed",
+					cspiItem.Name, cspcObj.Name)
+			}
+		}
+	}
+	return nil
+}
+
+// canPerformCSPICleanup performs the validation if the cleanup for the
+// CSPI can begin
+func canPerformCSPICleanup(cspiObj *apis.CStorPoolInstance) bool {
```

Comment:
this returns true even if CSPI just got created and so, it have CSPCFinalizer and doesn't have PoolProtectionFinalizer.
Should we also check for non-zero deletionTimestamp on CSPI?

Reply:
The first check in the for loop of cspiList is for DeletionTimestamp. We can move that check to a predicate function and add it to predicates.

Reply:
does it have race with the delete code in CSPI-mgmt?
zpool.Delete in destroy need to be called after checking for 'PoolProtectionFinalizer'.
And, BTW, when does pool pod goes away?

Reply:
There is no race as the cleanup will be skipped until the PoolProtectionFinalizer is removed. The pool pod goes away only after CSPI goes away because of OwnerRef of CSPI on pool deployment.

Reply:
got to see the check for DeletionTimestamp in the loop

Reply:
race is as follows:
in pool-mgmt, zpool.Delete in destroy can get called on every reconciliation if delete timestamp is non-zero on CSPI.
As there is another finalizer, CSPI won't go away until cspc-operator removes the other finalizer and pool-mgmt keep calling this zpool.Delete.
When that is going on (i..e, zpool.Delete is going on), cspc-operator might have removed the finalizer and the object goes away. This is what I'm calling as race.
This probably should be fine, but, want to understand when the pool pod goes away.

Reply:
The PoolProtectionFinalizer will be removed only after zpool.Delete is successful. After the PoolProtectionFinalizer is removed then only cspc-operator can remove the CSPCFinalizer.